File: benchmarks/script_entrypoints_test.py
import os
import subprocess
import sys
import unittest
from pathlib import Path

sys.dont_write_bytecode = True

import benchmark_discovery as discovery
import timing_compare


ROOT = Path(__file__).resolve().parents[1]


def module_path() -> str:
    for line in (ROOT / "go.mod").read_text().splitlines():
        if line.startswith("module "):
            return line.split()[1]
    raise AssertionError("go.mod has no module declaration")


class ScriptEntrypointConsistencyTest(unittest.TestCase):
    def test_diag_shell_uses_shared_discovery(self):
        diag = (ROOT / "scripts" / "diag.sh").read_text()
        self.assertIn("import benchmark_discovery as discovery", diag)
        self.assertIn("discovery.GROUPS", diag)
        self.assertIn("selector in discovery.GROUPS", diag)
        self.assertIn("discovery.resolve_script_path(root, selector)", diag)
        self.assertNotIn("domain_list_for()", diag)

    def test_benchmark_shell_wrappers_exec_matching_python(self):
        regression = (ROOT / "benchmarks" / "regression_guard.sh").read_text()
        self.assertIn('exec go run ./cmd/leia bench regression-guard "$@"', regression)
        strict = (ROOT / "benchmarks" / "strict_guard.sh").read_text()
        self.assertIn('exec go run ./cmd/leia bench strict "$@"', strict)

    def test_q_columnar_suite_wraps_timing_compare(self):
        wrapper = (ROOT / "benchmarks" / "q_columnar_suite.sh").read_text()
        self.assertIn("go run ./cmd/leia bench compare", wrapper)
        self.assertIn("--no-luajit", wrapper)
        for bench in (
            "data/q_columnar_eval_primitives",
            "data/q_columnar_qsql_filter_project",
            "data/q_columnar_qsql_group_xbar",
            "data/q_columnar_qsql_asof_join",
        ):
            with self.subTest(bench=bench):
                self.assertIn(f"--bench={bench}", wrapper)

    def test_scripts_performance_gate_wraps_benchmark_python_tools(self):
        gate = (ROOT / "scripts" / "performance_gate.sh").read_text()
        self.assertIn("go run ./cmd/leia bench compare", gate)
        self.assertIn("go run ./cmd/leia bench strict", gate)
        self.assertIn("--progress", gate)
        self.assertIn('--jobs="$JOBS"', gate)
        self.assertIn("table/table_field_access", gate)
        self.assertIn("--syntax-smoke", gate)
        self.assertIn("SYNTAX_SMOKE_BENCHES=(", gate)
        self.assertIn('PROFILE="syntax_smoke"', gate)
        self.assertIn('STRICT=0', gate)
        self.assertIn("--quick-phase-smoke", gate)
        self.assertIn('PROFILE="quick_phase_smoke"', gate)
        self.assertIn("STRICT_SMOKE_BENCHES=(", gate)
        self.assertIn('for bench in "${STRICT_SMOKE_BENCHES[@]}"; do', gate)
        self.assertIn('if [ "$PROFILE" = "full" ]; then', gate)
        self.assertIn('for bench in "${STRICT_CORE_BENCHES[@]}"; do', gate)

    def test_python_benchmark_entrypoints_share_discovery_groups(self):
        expected = tuple(discovery.GROUPS)
        self.assertEqual(tuple(timing_compare.GROUPS), expected)

    def test_release_scripts_gate_current_module_path(self):
        expected = module_path()
        for script in ("production_check.sh", "release_artifacts_check.sh"):
            with self.subTest(script=script):
                text = (ROOT / "scripts" / script).read_text()
                self.assertIn(expected, text)
                self.assertNotIn("github.com/leia/leia", text)
                self.assertNotIn("github.com/Never-Labs/leia", text)

    def test_production_check_full_plan_avoids_go_test_duplicates(self):
        text = (ROOT / "scripts" / "production_check.sh").read_text()
        full_plan = text.split("build_full_plan() {", 1)[1].split("\n}", 1)[0]

        self.assertIn('add_go_test "Correctness"', full_plan)
        self.assertIn('add_skip "Feature Matrix" "covered by Correctness', full_plan)
        self.assertIn('add_skip "Release Matrix Metadata" "covered by Correctness', full_plan)
        self.assertNotIn('add_go_test "Feature Matrix"', full_plan)
        self.assertNotIn('add_go_test "Release Matrix Metadata"', full_plan)

    def test_release_profile_requires_release_only_tool_smokes(self):
        production = (ROOT / "scripts" / "production_check.sh").read_text()
        distribution = (ROOT / "scripts" / "release_distribution_check.sh").read_text()
        ci = (ROOT / "cmd" / "leia" / "ci.go").read_text()

        self.assertIn("--release-profile", production)
        self.assertIn("scripts/run.sh editor --require-tree-sitter", production)
        self.assertIn("scripts/run.sh release-dist --require-goreleaser", production)
        self.assertIn("scripts/run.sh release-check", production)
        self.assertIn("--require-goreleaser", distribution)
        self.assertIn("goreleaser CLI is required for release distribution profile", distribution)
        self.assertIn('"--full", "--release-profile"', ci)

    def test_production_release_profile_list_includes_required_tool_flags(self):
        proc = subprocess.run(
            ["bash", "scripts/production_check.sh", "--full", "--release-profile", "--list"],
            cwd=ROOT,
            check=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            text=True,
        )

        self.assertIn("Release profile: critical release tool skips are treated as failures.", proc.stdout)
        self.assertIn("scripts/run.sh editor --require-tree-sitter", proc.stdout)
        self.assertIn("scripts/run.sh release-dist --require-goreleaser", proc.stdout)
        self.assertIn("scripts/run.sh release-check", proc.stdout)

    def test_release_distribution_require_goreleaser_fails_without_cli(self):
        env = os.environ.copy()
        env["PATH"] = "/usr/bin:/bin"
        if subprocess.run(
            ["bash", "-lc", "command -v goreleaser >/dev/null 2>&1"],
            env=env,
            check=False,
        ).returncode == 0:
            self.skipTest("goreleaser is available on the restricted PATH")

        proc = subprocess.run(
            ["bash", "scripts/release_distribution_check.sh", "--require-goreleaser"],
            cwd=ROOT,
            env=env,
            check=False,
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            text=True,
        )

        self.assertEqual(proc.returncode, 1, proc.stdout)
        self.assertIn("goreleaser CLI is required for release distribution profile", proc.stdout)

    def test_scripts_have_valid_type_specific_syntax(self):
        scripts = sorted(path for path in (ROOT / "scripts").iterdir() if path.is_file())
        self.assertGreater(len(scripts), 0)
        for script in scripts:
            with self.subTest(script=script.name):
                if script.suffix == ".sh":
                    cmd = ["bash", "-n", str(script)]
                elif script.suffix == ".py":
                    cmd = [
                        "python3",
                        "-c",
                        (
                            "import ast, pathlib, sys; "
                            "path = pathlib.Path(sys.argv[1]); "
                            "ast.parse(path.read_text(encoding='utf-8'), filename=str(path))"
                        ),
                        str(script),
                    ]
                elif script.suffix == ".leia":
                    cmd = ["go", "run", "./cmd/leia", "lint", str(script)]
                else:
                    self.fail(f"unsupported script entrypoint type: {script.name}")
                subprocess.run(
                    cmd,
                    cwd=ROOT,
                    check=True,
                    stdout=subprocess.PIPE,
                    stderr=subprocess.PIPE,
                    text=True,
                )


if __name__ == "__main__":
    unittest.main()

File: benchmarks/script_entrypoints_test.go
package benchmarks

import (
	"errors"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"
	"testing"
)

func repoRootForScriptEntrypoints(t *testing.T) string {
	t.Helper()
	dir, err := os.Getwd()
	if err != nil {
		t.Fatal(err)
	}
	return filepath.Dir(dir)
}

func readRepoFile(t *testing.T, root string, parts ...string) string {
	t.Helper()
	data, err := os.ReadFile(filepath.Join(append([]string{root}, parts...)...))
	if err != nil {
		t.Fatal(err)
	}
	return string(data)
}

func requireContains(t *testing.T, text, want string) {
	t.Helper()
	if !strings.Contains(text, want) {
		t.Fatalf("text missing %q", want)
	}
}

func requireNotContains(t *testing.T, text, want string) {
	t.Helper()
	if strings.Contains(text, want) {
		t.Fatalf("text unexpectedly contains %q", want)
	}
}

func modulePathForScriptEntrypoints(t *testing.T, root string) string {
	t.Helper()
	for _, line := range strings.Split(readRepoFile(t, root, "go.mod"), "\n") {
		if strings.HasPrefix(line, "module ") {
			return strings.Fields(line)[1]
		}
	}
	t.Fatal("go.mod has no module declaration")
	return ""
}

func runScriptEntrypointCommand(t *testing.T, root string, env []string, args ...string) (string, int) {
	t.Helper()
	cmd := exec.Command(args[0], args[1:]...)
	cmd.Dir = root
	if env != nil {
		cmd.Env = env
	}
	out, err := cmd.CombinedOutput()
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			return string(out), exitErr.ExitCode()
		}
		t.Fatalf("%v: %v\n%s", args, err, out)
	}
	return string(out), 0
}

func TestDiagShellUsesSharedDiscovery(t *testing.T) {
	root := repoRootForScriptEntrypoints(t)
	diag := readRepoFile(t, root, "scripts", "diag.sh")
	requireContains(t, diag, "DOMAIN_GROUPS=(numeric recursion table calls string concurrency data app control precision)")
	requireContains(t, diag, "DEFAULT_ORDER=(")
	requireContains(t, diag, "add_group_benches()")
	requireContains(t, diag, "resolve_selector()")
	requireContains(t, diag, `find "$bench_dir" -maxdepth 1 -type f -name '*.leia' | sort`)
	requireNotContains(t, diag, "import benchmark_discovery as discovery")
	requireNotContains(t, diag, "domain_list_for()")
}

func TestBenchmarkShellWrappersExecLeiaCLI(t *testing.T) {
	root := repoRootForScriptEntrypoints(t)
	requireContains(t, readRepoFile(t, root, "benchmarks", "regression_guard.sh"), `exec go run ./cmd/leia bench regression-guard "$@"`)
	requireContains(t, readRepoFile(t, root, "benchmarks", "strict_guard.sh"), `exec go run ./cmd/leia bench strict "$@"`)
}

func TestQColumnarSuiteWrapsLeiaBenchCompare(t *testing.T) {
	root := repoRootForScriptEntrypoints(t)
	wrapper := readRepoFile(t, root, "benchmarks", "q_columnar_suite.sh")
	requireContains(t, wrapper, "go run ./cmd/leia bench compare")
	requireContains(t, wrapper, "--no-luajit")
	for _, bench := range []string{
		"data/q_columnar_eval_primitives",
		"data/q_columnar_qsql_filter_project",
		"data/q_columnar_qsql_group_xbar",
		"data/q_columnar_qsql_asof_join",
	} {
		requireContains(t, wrapper, "--bench="+bench)
	}
}

func TestScriptsPerformanceGateWrapsLeiaBenchTools(t *testing.T) {
	root := repoRootForScriptEntrypoints(t)
	gate := readRepoFile(t, root, "scripts", "performance_gate.sh")
	for _, want := range []string{
		"go run ./cmd/leia bench compare",
		"go run ./cmd/leia bench strict",
		"--progress",
		`--jobs="$JOBS"`,
		"table/table_field_access",
		"--syntax-smoke",
		"SYNTAX_SMOKE_BENCHES=(",
		`PROFILE="syntax_smoke"`,
		"STRICT=0",
		"--quick-phase-smoke",
		`PROFILE="quick_phase_smoke"`,
		"STRICT_SMOKE_BENCHES=(",
		`for bench in "${STRICT_SMOKE_BENCHES[@]}"; do`,
		`if [ "$PROFILE" = "full" ] || [ "$PROFILE" = "release" ]; then`,
		`PROFILE="release"`,
		`MIN_SAMPLE_SECONDS=0.050`,
		`MAX_REPEAT=64`,
		`HEAD_REF=""`,
		`for bench in "${STRICT_CORE_BENCHES[@]}"; do`,
	} {
		requireContains(t, gate, want)
	}
}

func TestReleaseScriptsGateCurrentModulePath(t *testing.T) {
	root := repoRootForScriptEntrypoints(t)
	expected := modulePathForScriptEntrypoints(t, root)
	for _, script := range []string{"production_check.sh", "release_artifacts_check.sh"} {
		text := readRepoFile(t, root, "scripts", script)
		requireContains(t, text, expected)
		requireNotContains(t, text, "github.com/leia/leia")
		requireNotContains(t, text, "github.com/Never-Labs/leia")
	}
}

func TestProductionCheckFullPlanAvoidsGoTestDuplicates(t *testing.T) {
	root := repoRootForScriptEntrypoints(t)
	text := readRepoFile(t, root, "scripts", "production_check.sh")
	parts := strings.SplitN(text, "build_full_plan() {", 2)
	if len(parts) != 2 {
		t.Fatal("build_full_plan not found")
	}
	fullPlan := strings.SplitN(parts[1], "\n}", 2)[0]
	requireContains(t, fullPlan, `add_go_test "Correctness"`)
	requireContains(t, fullPlan, `add_skip "Feature Matrix" "covered by Correctness`)
	requireContains(t, fullPlan, `add_skip "Release Matrix Metadata" "covered by Correctness`)
	requireNotContains(t, fullPlan, `add_go_test "Feature Matrix"`)
	requireNotContains(t, fullPlan, `add_go_test "Release Matrix Metadata"`)
}

func TestReleaseProfileRequiresReleaseOnlyToolSmokes(t *testing.T) {
	root := repoRootForScriptEntrypoints(t)
	production := readRepoFile(t, root, "scripts", "production_check.sh")
	distribution := readRepoFile(t, root, "scripts", "release_distribution_check.sh")
	ci := readRepoFile(t, root, "cmd", "leia", "ci.go")
	for _, want := range []string{
		"--release-profile",
		"scripts/run.sh editor --require-tree-sitter",
		"scripts/run.sh release-dist --require-goreleaser",
		"scripts/run.sh release-notes-gate",
		"scripts/run.sh release-artifacts-gate",
	} {
		requireContains(t, production, want)
	}
	requireContains(t, distribution, "--require-goreleaser")
	requireContains(t, distribution, "github.com/goreleaser/goreleaser/v2@v2.16.0")
	requireContains(t, distribution, "pinned go-run fallback")
	requireContains(t, ci, `"--full", "--release-profile"`)
}

func TestProductionReleaseProfileListIncludesRequiredToolFlags(t *testing.T) {
	root := repoRootForScriptEntrypoints(t)
	out, code := runScriptEntrypointCommand(t, root, nil, "bash", "scripts/production_check.sh", "--full", "--release-profile", "--list")
	if code != 0 {
		t.Fatalf("production_check exit %d:\n%s", code, out)
	}
	for _, want := range []string{
		"Release profile: critical release tool skips are treated as failures.",
		"scripts/run.sh editor --require-tree-sitter",
		"scripts/run.sh release-dist --require-goreleaser",
		"scripts/run.sh release-notes-gate",
		"scripts/run.sh release-artifacts-gate",
	} {
		requireContains(t, out, want)
	}
}

func TestReleaseDistributionRequireGoreleaserFailsWithoutCLIOrGo(t *testing.T) {
	root := repoRootForScriptEntrypoints(t)
	tmpBin := t.TempDir()
	for _, name := range []string{"awk", "bash", "chmod", "cp", "dirname", "env", "grep", "install", "mkdir", "mktemp", "openssl", "rm", "sh", "shasum", "sha256sum", "tar", "unzip", "zip"} {
		path, err := exec.LookPath(name)
		if err != nil {
			continue
		}
		if err := os.Symlink(path, filepath.Join(tmpBin, name)); err != nil && !errors.Is(err, os.ErrExist) {
			t.Fatalf("symlink %s into test PATH: %v", name, err)
		}
	}
	if _, err := os.Stat(filepath.Join(tmpBin, "bash")); err != nil {
		t.Fatalf("test PATH needs bash: %v", err)
	}
	env := append(os.Environ(), "PATH="+tmpBin)
	out, code := runScriptEntrypointCommand(t, root, env, "bash", "scripts/release_distribution_check.sh", "--require-goreleaser")
	if code != 1 {
		t.Fatalf("release_distribution_check exit %d, want 1:\n%s", code, out)
	}
	requireContains(t, out, "goreleaser check requires a local goreleaser CLI or go for pinned go-run fallback")
}

func TestScriptsHaveValidTypeSpecificSyntax(t *testing.T) {
	root := repoRootForScriptEntrypoints(t)
	entries, err := os.ReadDir(filepath.Join(root, "scripts"))
	if err != nil {
		t.Fatal(err)
	}
	sort.Slice(entries, func(i, j int) bool { return entries[i].Name() < entries[j].Name() })
	if len(entries) == 0 {
		t.Fatal("scripts directory is empty")
	}
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		name := entry.Name()
		path := filepath.Join("scripts", name)
		var args []string
		switch filepath.Ext(name) {
		case ".sh":
			args = []string{"bash", "-n", path}
		case ".leia":
			args = []string{"go", "run", "./cmd/leia", "lint", path}
		default:
			t.Fatalf("unsupported script entrypoint type: %s", name)
		}
		if out, code := runScriptEntrypointCommand(t, root, nil, args...); code != 0 {
			t.Fatalf("%s syntax exit %d:\n%s", name, code, out)
		}
	}
}
